test/integration/util/shell.go

```diff
@@ -0,0 +1,282 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+
+	"github.com/DATA-DOG/godog/gherkin"
+	"github.com/minishift/minishift/pkg/util/os"
+)
+
+const (
+	exitCodeIdentifier = "exitCodeOfLastCommandInShell="
+
+	bashExitCodeCheck       = "echo %v$?"
+	fishExitCodeCheck       = "echo %v$status"
+	tcshExitCodeCheck       = "echo %v$?"
+	zshExitCodeCheck        = "echo %v$?"
+	cmdExitCodeCheck        = "echo %v%%errorlevel%%"
+	powershellExitCodeCheck = "echo %v$lastexitcode"
+)
+
+var (
+	shell shellInstance
+)
+
+type shellInstance struct {
+	startArgument    []string
+	name             string
+	checkExitCodeCmd string
+
+	minishiftPath string
+
+	instance *exec.Cmd
+	outbuf   bytes.Buffer
+	errbuf   bytes.Buffer
+	excbuf   bytes.Buffer
+
+	outPipe io.ReadCloser
+	errPipe io.ReadCloser
+	inPipe  io.WriteCloser
+
+	outScanner *bufio.Scanner
+	errScanner *bufio.Scanner
+
+	stdoutChannel   chan string
+	stderrChannel   chan string
+	exitCodeChannel chan string
+}
+
+func SetShell(name string, mpath string) {
+	shell.name = name
+	shell.minishiftPath = mpath
+}
+
+func configureShell() {
+	switch shell.name {
+	case "bash":
+		shell.checkExitCodeCmd = fmt.Sprintf(bashExitCodeCheck, exitCodeIdentifier)
+	case "tcsh":
+		shell.checkExitCodeCmd = fmt.Sprintf(tcshExitCodeCheck, exitCodeIdentifier)
+	case "zsh":
+		shell.checkExitCodeCmd = fmt.Sprintf(zshExitCodeCheck, exitCodeIdentifier)
+	case "cmd":
+		shell.checkExitCodeCmd = fmt.Sprintf(cmdExitCodeCheck, exitCodeIdentifier)
+	case "powershell":
+		shell.startArgument = []string{"-Command", "-"}
+		shell.checkExitCodeCmd = fmt.Sprintf(powershellExitCodeCheck, exitCodeIdentifier)
+	case "fish":
+		fmt.Println("Fish shell is currently not supported by integration tests. Default shell for the OS will be used.")
+		shell.name = ""
+		fallthrough
+	default:
+		if shell.name != "" {
+			fmt.Printf("Shell %v is not supported, will set the default shell for the OS to be used.\n", shell.name)
+		}
+		switch os.CurrentOS() {
+		case "darwin", "linux":
+			shell.name = "bash"
+			shell.checkExitCodeCmd = fmt.Sprintf(bashExitCodeCheck, exitCodeIdentifier)
+		case "windows":
+			shell.name = "powershell"
+			shell.startArgument = []string{"-Command", "-"}
+			shell.checkExitCodeCmd = fmt.Sprintf(powershellExitCodeCheck, exitCodeIdentifier)
+		}
+	}
+
+	return
+}
+
+func StartShellInstance() error {
+	var err error
+	shell.stdoutChannel = make(chan string)
```

Comment:
You really should have a constructor which creates the instance. Now you create an empty `shell shellInstance` as a var and keep setting various bits and pieces. That's marginally better. However, the idea behind a struct is that you control its initalization and only make available what needs to be made available. Also you guarantee that if you are dealing with an instance of shellInstance it is in a consistent state. See also http://www.golangpatterns.info/object-oriented/constructors

note constructors do not really exist in Golang. We are talking factory methods really. Check our codebase as well. There you see it at play at several places as well.